Reworked plugin queues

hanshuebner opened the pull request:
### Description

This PR contains the documentation updates for the reworked plugin queues.

https://github.com/Kong/kong/pull/10172

### Testing instructions

Netlify link: https://deploy-preview-5088--kongdocs.netlify.app/


### Checklist 

- [X] Review label added <!-- (see below) -->
- [X] PR pointed to correct branch (`main` for immediate publishing, or a release branch: e.g. `release/gateway-3.2`, `release/deck-1.17`)

acgoldsmith commented:
@hanshuebner , I'll work with you on this for now. We haven't assigned a writer yet for this piece of 3.3 work.

hanshuebner commented:
> @hanshuebner When you have the time, could you change the base branch to `release/gateway-3.3` instead of `main`?

Done.  I'll add a couple of things and resolve the conflicts by the end of the week.

hanshuebner commented:
@cloudjumpercat I have updated the parameter documentation and added links to the common queue configuration parameters from each of the plugins.  I am not sure, however, if I did it right.  Also, links back to the documentation should be added.  Can you have a look and let me know if anything is missing from my end?

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -142,10 +150,24 @@ params:
 
+## Queueing
+
+The http-log plugin uses internal queues to decouple the production of
```
```suggestion
The HTTP Log plugin uses internal queues to decouple the production of
```

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -142,10 +150,24 @@ params:
+
+The http-log plugin uses internal queues to decouple the production of
+log entries from their transmission to the upstream log server.  In
+contrast to other plugins that uses queues, it shares one queue
```
```suggestion
contrast to other plugins that use queues, it shares one queue
```

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -142,10 +150,24 @@ params:
+log entries from their transmission to the upstream log server.  In
+contrast to other plugins that uses queues, it shares one queue
+between all plugin instances that use the same log server parameter.
+Equivalence of the log server is determined by the parameters
+`http_endpoint`, `method`, `content_type`, `timeout` and `keepalive`.
```
```suggestion
The equivalence of the log server is determined by the parameters
`http_endpoint`, `method`, `content_type`, `timeout`, and `keepalive`.
```

cloudjumpercat commented on app/_includes/md/plugins-hub/queue-parameters.md:
```diff
@@ -0,0 +1,62 @@
+made to the upstream server under high load situations and provides
+for buffering during temporary network or upstream outages.
+
+A number of parameters can be tuned to influence the behavior and
+capacity of the queues used by the plugin.  Please see [XXX](the
+documentation on queues) for detail information on the usage of these
+parameters.
```
```suggestion
You can set several parameters to configure the behavior and
capacity of the queues used by the plugin. For more information about how to use these parameters, see [Batch Queuing Reference](gateway/latest/kong-plugins/batch-queue/reference/) in the {{site.base_gateway}} documentation.
```
I think that's the link you were looking for?

cloudjumpercat commented on app/_includes/md/plugins-hub/queue-parameters.md:
```diff
@@ -0,0 +1,62 @@
+The queue parameters all reside in a record under the key `queue` in
+the `config` parameter section of the plugin.
+
+parameters:
```
Did you want this line and below to display like the table like the rest of the parameters? I think it would have to live in the same spot as those other parameters (I don't think it will render in this section).

hanshuebner replied:
Can we somehow make the queue parameters into a separate table for each plugin?

cloudjumpercat replied:
@hanshuebner I asked about making the parameters into a separate table and it's not possible. The queue parameters have to be included in each plugin’s index file manually as part of the params section.

cloudjumpercat commented on app/_src/gateway/production/sizing-guidelines.md:
```diff
@@ -93,6 +93,43 @@ around **500MB** of memory allocated per worker process.
 
+### Plugin queues
+
+Several plugins that are distributed with {{site.base_gateway}}
+utilize internal, in-memory queues to decouple production of data from
+the transmission to an upstream server.  These queues reduce the
+number of concurrent requests that are made to an upstream server
+under high load conditions and provide for buffering during temporary
+network and upstream outages.  Please see the [Plugin Queues]()
+documentation for details on {{site.base_gateway}}'s internal
+queueing system.
```
```suggestion
Several plugins that are distributed with {{site.base_gateway}}
utilize internal, in-memory queues to decouple production of data from
the transmission to an upstream server. These queues reduce the
number of concurrent requests that are made to an upstream server
under high load conditions and provide for buffering during temporary
network and upstream outages. For more information about {{site.base_gateway}}'s internal queueing system, see [About Batch Queuing](/gateway/{{page.kong_version}}/kong-plugins/batch-queue/).
```

cloudjumpercat commented on app/_src/gateway/production/sizing-guidelines.md:
```diff
@@ -93,6 +93,43 @@ around **500MB** of memory allocated per worker process.
+they can hold in terms of their capacity configuration.
+
+Most plugins use one queue per plugin instance, with the exception of
+the http-log plugin, which uses one queue per log server upstream
```
```suggestion
the HTTP Log plugin, which uses one queue per log server upstream
```

cloudjumpercat commented on app/_src/gateway/production/sizing-guidelines.md:
```diff
@@ -93,6 +93,43 @@ around **500MB** of memory allocated per worker process.
+The `queue.max_entries` configuration parameter determines how many
+entries can be waiting for transmission on a given queue.  Once this
+limit is reached, the oldest entry is removed when a new entry is
+enqueued.  While it is not possible to precisely predict how much
```
```suggestion
enqueued. While it isn't possible to precisely predict how much
```

Guaris replied:
```suggestion
enqueued. While it is not possible to precisely predict how much
```

cloudjumpercat commented on app/_src/gateway/kong-plugins/batch-queue/index.md:
```diff
@@ -0,0 +1,67 @@
+Previously, the queue didn't have any capacity limits. This resulted in a worker process exhausting all available memory if data was produced fast enough and the upstream server would be slow or unavailable.
+
+Now, you can configure the maximum number of entries that can be queued at any time with the `max_entries` queue parameter. When a queue reaches the maximum number of entries queued and another entry is enqueued, the oldest entry in the queue is deleted to make space for the new entry.
+The queue code provides warning log entries when it reaches a capacity threshold of 80% and when it starts to delete entries from the queue. It also writes log entries when the situation normalizes.
```
```suggestion
The queue code provides warning log entries when it reaches a capacity threshold of 80% and when it starts to delete entries from the queue. It also writes log entries when the situation normalizes.

For more information about queue size limits, see the [plugin queue resource sizing guidelines](/gateway/{{page.kong_version}}/production/sizing-guidelines/#plugin-queues).
```

cloudjumpercat commented on app/_src/gateway/kong-plugins/batch-queue/reference.md:
```diff
@@ -0,0 +1,30 @@
+
+## More information
+
+[About Batch Queuing](/gateway/{{page.kong_version}}/kong-plugins/batch-queue/) - Find out how batch queuing works.
```
```suggestion
* [About Batch Queuing](/gateway/{{page.kong_version}}/kong-plugins/batch-queue/) - Find out how batch queuing works.
* [Plugin queue resource sizing guidelines](/gateway/{{page.kong_version}}/production/sizing-guidelines/#plugin-queues) - Learn about queue size limits. 
```

cloudjumpercat reviewed (commented):
Left a few suggestions and a couple questions for you on some of the docs I wrote. Let me know if you have any questions! Things are looking pretty good so far!

cloudjumpercat commented on app/_hub/kong-inc/datadog/_index.md:
```diff
@@ -86,19 +86,33 @@ params:
+    - name: queue
+      required: false
+      datatype: record
+      description: Configuration parameters for queue (XXX link to queue parameters missing)
```
```suggestion
      description: |
  Configuration parameters for batch queuing. For more information, see [Queuing](#queueing).
```

cloudjumpercat replied:
I _think_ this will make the link work.

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -89,17 +89,25 @@ params:
+    - name: queue
+      required: false
+      datatype: record
+      description: Configuration parameters for queue (XXX link to queue parameters missing)
```
```suggestion
      description: |
  Configuration parameters for batch queuing. For more information, see [Queuing](#queueing).
```

cloudjumpercat commented on app/_hub/kong-inc/opentelemetry/_index.md:
```diff
@@ -34,6 +34,12 @@ params:
+    - name: queue
+      required: false
+      datatype: record
+      description: Configuration parameters for queue (XXX link to queue parameters missing)
```
```suggestion
      description: |
  Configuration parameters for batch queuing. For more information, see [Queuing](#queueing).
```

cloudjumpercat commented on app/_hub/kong-inc/statsd/_index.md:
```diff
@@ -115,19 +115,27 @@ params:
+    - name: queue
+      required: false
+      datatype: record
+      description: Configuration parameters for queue (XXX link to queue parameters missing)
```
```suggestion
      description: |
  Configuration parameters for batch queuing. For more information, see [Queuing](#queueing).
```

cloudjumpercat commented on app/_hub/kong-inc/zipkin/_index.md:
```diff
@@ -81,6 +81,11 @@ params:
+    - name: queue
+      required: false
+      datatype: record
+      description: Configuration parameters for queue (XXX link to queue parameters missing)
```
```suggestion
      description: |
  Configuration parameters for batch queuing. For more information, see [Queuing](#queueing).
```

hanshuebner commented on app/_src/gateway/kong-plugins/batch-queue/index.md:
```diff
@@ -0,0 +1,67 @@
+
+Some analytic and monitoring plugins, like HTTP Log and Datadog, must send request information that was processed by {{site.base_gateway}} to another server (the upstream server). 
+
+Without batch queuing, the information is during the log phase to avoid introducing any latency to request or response. This isn't ideal because directly sending data from the log handler to the upstream server creates a large number of requests when {{site.base_gateway}} is under a high load. This results in ....
```
Something like:

> Sending out a large number of concurrent outbound requests to the log server potentially overloads that server.  Also, the resource usage of these requests on the machine running {{site.base_gateway}} can negatively affects is ability to proxy requests.

hanshuebner commented on app/_src/gateway/kong-plugins/batch-queue/index.md:
```diff
@@ -0,0 +1,67 @@
+
+When {{site.base_gateway}} processes a request from certain plugins, it uses batch queuing to group the requests togehter before sending them to the upstream server. 
+
+[DIAGRAM HERE?]
```
Here's a link to the slides: https://docs.google.com/presentation/d/1oIKuHTJkCxUlOdz1pV5nJtoQM0LQ8JI0EzjmU_a5PXA/edit?usp=sharing
Here's the diagrams for completeness:
https://lucid.app/lucidchart/invitations/accept/inv_88649aa5-7b7d-4383-8823-3b826905ced2

cloudjumpercat commented:
@hanshuebner I think all we need is to add the queue params to the index files of the plugins (with the rest of the other parameters), I requested a diagram from Marketing but I can always comment it out and we can publish this without the diagram and add it back in later, and I'll also ask a writer to look over the content I wrote. Then it should be good to merge!

cloudjumpercat commented:
@lena-larionova or @Guaris If either of you have the time, could you review just the pages I wrote? Thanks!

https://deploy-preview-5088--kongdocs.netlify.app/gateway/latest/kong-plugins/batch-queue/
https://deploy-preview-5088--kongdocs.netlify.app/gateway/latest/kong-plugins/batch-queue/reference/

hanshuebner commented:
@cloudjumpercat I have spliced the `queue-parameters.md` file into the documentation of all the plugins that support queues and removed it.

I saw that you are using the term "batch queueing", which I don't think we should do.  I'd prefer if we'd talk about just "queueing", as batching is something that is optional and "batch queue" is a term that has a different meaning in the software industry.

The links to the reference document don't work, possibly because they include `gateway/latest/` as a relative prefix.  Is that intentional?

cloudjumpercat commented on app/_hub/kong-inc/datadog/_index.md:
```diff
@@ -87,19 +87,87 @@ params:
+
+The queue parameters all reside in a record under the key `queue` in
+the `config` parameter section of the plugin.
+
```
```suggestion
{% endif_version %}
```

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -142,10 +191,22 @@ params:
       minimum_version: "2.4.x"
 ---
 
+## Queueing
```
```suggestion
{% if_version gte:3.3.x %}
## Queueing
```

cloudjumpercat commented on app/_hub/kong-inc/opentelemetry/_index.md:
```diff
@@ -107,10 +113,70 @@ params:
 
-The OpenTelemetry plugin is built on top of {{site.base_gateway}}'s tracing API
-and is intended to be fully compatible with the OpenTelemetry specification.
+## Queueing
```
```suggestion
{% if_version gte:3.3.x %}
## Queueing
```

cloudjumpercat commented on app/_hub/kong-inc/statsd/_index.md:
```diff
@@ -139,6 +188,25 @@ params:
     65507 according to UDP protocol. Consider the MTU of the network when setting this parameter.
 ---
 
+## Queueing
```
```suggestion
{% if_version gte:3.3.x %}
## Queueing
```

cloudjumpercat commented on app/_hub/kong-inc/zipkin/_index.md:
```diff
@@ -248,6 +294,22 @@ params:
         * `tags`: Include the duration of each phase as a tag.
 
+---
+## Queueing
```
```suggestion

{% if_version gte:3.3.x %}
## Queueing
```

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -142,10 +191,22 @@ params:
+The equivalence of the log server is determined by the parameters
+`http_endpoint`, `method`, `content_type`, `timeout`, and `keepalive`.
+All plugin instances that have the same values for these parameters
+share one queue.
```
```suggestion
share one queue.
{% endif_version %}
```

cloudjumpercat commented on app/_hub/kong-inc/statsd/_index.md:
```diff
@@ -139,6 +188,25 @@ params:
+in the {{site.base_gateway}} documentation.
+
+The queue parameters all reside in a record under the key `queue` in
+the `config` parameter section of the plugin.
```
```suggestion
the `config` parameter section of the plugin.
{% endif_version %}
```

cloudjumpercat commented on app/_hub/kong-inc/zipkin/_index.md:
```diff
@@ -248,6 +294,22 @@ params:
+in the {{site.base_gateway}} documentation.
+
+The queue parameters all reside in a record under the key `queue` in
+the `config` parameter section of the plugin.
```
```suggestion
the `config` parameter section of the plugin.
{% endif_version %}
```

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -142,10 +191,22 @@ params:
+
 ## Log format
 
 {:.note}
```
```suggestion
{% if_version lte:3.2.x %}
{:.note}
> **Note:** If the `queue_size` argument > 1, a request is logged as an array of JSON objects.
{% endif_version %}
{% if_version gte:3.3.x %}
{:.note}
```

cloudjumpercat commented on app/_hub/kong-inc/datadog/_index.md:
```diff
@@ -87,19 +87,88 @@ params:
+          after the first entry was queued before the queue starts
+          processing entries.  This parameter has no effect when
+          `batch_max_size` is one because queued entries will be sent
+          immediately in that case.
```
```suggestion
          immediately in that case.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/datadog/_index.md:
```diff
@@ -87,19 +87,88 @@ params:
+      description: >
+          Maximum number of entries that can be waiting on the queue.
+          Once this number of requests is reached, the oldest entry is
+          deleted from the queue before a new one is added.
```
```suggestion
          deleted from the queue before a new one is added.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/datadog/_index.md:
```diff
@@ -87,19 +87,88 @@ params:
+          Maximum number of bytes that can be waiting on a queue.
+          Once this many bytes are present on a queue, old entries
+          up to the size of a new entry to be enqueued are deleted
+          from the queue.
```
```suggestion
          from the queue.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/datadog/_index.md:
```diff
@@ -87,19 +87,88 @@ params:
+          Maximum time (in seconds) between retries sending a batch of
+          entries. The interval between retries follows an
+          exponential back-off algorithm capped at this number of
+          seconds.
```
```suggestion
          seconds.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -89,17 +89,66 @@ params:
+      type: number
+      default: 1
+      description: Maximum number of entries to be processed together
+          as a batch.
```
```suggestion
          as a batch.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -89,17 +89,66 @@ params:
+          after the first entry was queued before the queue starts
+          processing entries.  This parameter has no effect when
+          `batch_max_size` is one because queued entries will be sent
+          immediately in that case.
```
```suggestion
          immediately in that case.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -89,17 +89,66 @@ params:
+      description: >
+          Maximum number of entries that can be waiting on the queue.
+          Once this number of requests is reached, the oldest entry is
+          deleted from the queue before a new one is added.
```
```suggestion
          deleted from the queue before a new one is added.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -89,17 +89,66 @@ params:
+          Maximum number of bytes that can be waiting on a queue.
+          Once this many bytes are present on a queue, old entries
+          up to the size of a new entry to be enqueued are deleted
+          from the queue.
```
```suggestion
          from the queue.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -89,17 +89,66 @@ params:
+          batch of entries.  Once this time is exceeded for a batch,
+          it is deleted from the queue without being sent.  If
+          this parameter is set to -1, no retries will be made for a
+          failed batch.
```
```suggestion
          failed batch.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/http-log/_index.md:
```diff
@@ -89,17 +89,66 @@ params:
+          Maximum time (in seconds) between retries sending a batch of
+          entries. The interval between retries follows an
+          exponential back-off algorithm capped at this number of
+          seconds.
```
```suggestion
          seconds.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/opentelemetry/_index.md:
```diff
@@ -107,10 +113,72 @@ params:
+      description: >
+          Maximum number of entries that can be waiting on the queue.
+          Once this number of requests is reached, the oldest entry is
+          deleted from the queue before a new one is added.
```
```suggestion
          deleted from the queue before a new one is added.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/opentelemetry/_index.md:
```diff
@@ -107,10 +113,72 @@ params:
+          Maximum number of bytes that can be waiting on a queue.
+          Once this many bytes are present on a queue, old entries
+          up to the size of a new entry to be enqueued are deleted
+          from the queue.
```
```suggestion
          from the queue.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/opentelemetry/_index.md:
```diff
@@ -107,10 +113,72 @@ params:
+          batch of entries.  Once this time is exceeded for a batch,
+          it is deleted from the queue without being sent.  If
+          this parameter is set to -1, no retries will be made for a
+          failed batch.
```
```suggestion
          failed batch.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/opentelemetry/_index.md:
```diff
@@ -107,10 +113,72 @@ params:
+          Maximum time (in seconds) between retries sending a batch of
+          entries. The interval between retries follows an
+          exponential back-off algorithm capped at this number of
+          seconds.
```
```suggestion
          seconds.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/statsd/_index.md:
```diff
@@ -115,19 +115,68 @@ params:
+      type: number
+      default: 1
+      description: Maximum number of entries to be processed together
+          as a batch.
```
```suggestion
          as a batch.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/statsd/_index.md:
```diff
@@ -115,19 +115,68 @@ params:
+          after the first entry was queued before the queue starts
+          processing entries.  This parameter has no effect when
+          `batch_max_size` is one because queued entries will be sent
+          immediately in that case.
```
```suggestion
          immediately in that case.
          minimum_version: "3.3.x"
```

cloudjumpercat replied:
```suggestion
          immediately in that case.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/statsd/_index.md:
```diff
@@ -115,19 +115,68 @@ params:
+      description: >
+          Maximum number of entries that can be waiting on the queue.
+          Once this number of requests is reached, the oldest entry is
+          deleted from the queue before a new one is added.
```
```suggestion
          deleted from the queue before a new one is added.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/statsd/_index.md:
```diff
@@ -115,19 +115,68 @@ params:
+          Maximum number of bytes that can be waiting on a queue.
+          Once this many bytes are present on a queue, old entries
+          up to the size of a new entry to be enqueued are deleted
+          from the queue.
```
```suggestion
          from the queue.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/statsd/_index.md:
```diff
@@ -115,19 +115,68 @@ params:
+          batch of entries.  Once this time is exceeded for a batch,
+          it is deleted from the queue without being sent.  If
+          this parameter is set to -1, no retries will be made for a
+          failed batch.
```
```suggestion
          failed batch.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/statsd/_index.md:
```diff
@@ -115,19 +115,68 @@ params:
+          Maximum time (in seconds) between retries sending a batch of
+          entries. The interval between retries follows an
+          exponential back-off algorithm capped at this number of
+          seconds.
```
```suggestion
          seconds.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/zipkin/_index.md:
```diff
@@ -81,6 +81,52 @@ params:
+      type: number
+      default: 1
+      description: Maximum number of entries to be processed together
+          as a batch.
```
```suggestion
          as a batch.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/zipkin/_index.md:
```diff
@@ -81,6 +81,52 @@ params:
+          after the first entry was queued before the queue starts
+          processing entries.  This parameter has no effect when
+          `batch_max_size` is one because queued entries will be sent
+          immediately in that case.
```
```suggestion
          immediately in that case.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/zipkin/_index.md:
```diff
@@ -81,6 +81,52 @@ params:
+          Maximum number of bytes that can be waiting on a queue.
+          Once this many bytes are present on a queue, old entries
+          up to the size of a new entry to be enqueued are deleted
+          from the queue.
```
```suggestion
          from the queue.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_hub/kong-inc/zipkin/_index.md:
```diff
@@ -81,6 +81,52 @@ params:
+          batch of entries.  Once this time is exceeded for a batch,
+          it is deleted from the queue without being sent.  If
+          this parameter is set to -1, no retries will be made for a
+          failed batch.
```
```suggestion
          failed batch.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_includes/md/plugins-hub/queue-parameters.md:
```diff
@@ -0,0 +1,47 @@
+          Maximum number of bytes that can be waiting on a queue.
+          Once this many bytes are present on a queue, old entries
+          up to the size of a new entry to be enqueued are deleted
+          from the queue.
```
```suggestion
          from the queue.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_includes/md/plugins-hub/queue-parameters.md:
```diff
@@ -0,0 +1,47 @@
+          batch of entries.  Once this time is exceeded for a batch,
+          it is deleted from the queue without being sent.  If
+          this parameter is set to -1, no retries will be made for a
+          failed batch.
```
```suggestion
          failed batch.
          minimum_version: "3.3.x"
```

cloudjumpercat commented on app/_includes/md/plugins-hub/queue-parameters.md:
```diff
@@ -0,0 +1,47 @@
+          Maximum time (in seconds) between retries sending a batch of
+          entries. The interval between retries follows an
+          exponential back-off algorithm capped at this number of
+          seconds.
```
```suggestion
          seconds.
          minimum_version: "3.3.x"
```

cloudjumpercat commented:
I think spacing errors are causing smoke test failures, I'll fix them in a bit.

Guaris commented on app/_src/gateway/kong-plugins/queue/index.md:
```diff
@@ -0,0 +1,64 @@
+
+## Why are queues needed? 
+
+Some analytic and monitoring plugins, like HTTP Log and Datadog, must send request information that was processed by {{site.base_gateway}} to another server (the upstream server). 
```
```suggestion
Some analytic and monitoring plugins, like HTTP Log and Datadog, must send request information that was processed by {{site.base_gateway}} to the upstream server.
```

Guaris commented on app/_src/gateway/kong-plugins/queue/index.md:
```diff
@@ -0,0 +1,64 @@
+
+Some analytic and monitoring plugins, like HTTP Log and Datadog, must send request information that was processed by {{site.base_gateway}} to another server (the upstream server). 
+
+Without queuing, the information is sent during the log phase to avoid introducing any latency to request or response. This isn't ideal because directly sending data from the log handler to the upstream server creates a large number of requests when {{site.base_gateway}} is under a high load. Sending out a large number of concurrent outbound requests can overload the log server and can negatively ...
```
```suggestion
Sending information directly from the log handler to the upstream server, instead of queuing it, avoids introducing latency to requests and responses. However, this approach can create a large number of requests when {{site.base_gateway}} is under a high load, which can overload the log server and negatively impact the ability of the machine running {{site.base_gateway}} to proxy requests.
```

Guaris commented on app/_src/gateway/kong-plugins/queue/index.md:
```diff
@@ -0,0 +1,64 @@
+The solution to this is to use batch requests. With queuing, request information is put in a configurable queue before being sent to the upstream server. This approach has the following benefits:
+* Reduces any possible concurrency on the upstream server
+* Helps deal with temporary outages of the upstream server due to network or administrative changes
+* Can reduce resource usage both in {{site.base_gateway}} and in the upstream server by collecting multiple entries from the queue in one request
```
```suggestion
* Can reduce resource usage both in {{site.base_gateway}} and on the upstream server by collecting multiple entries from the queue in one request
```

Guaris commented on app/_src/gateway/kong-plugins/queue/index.md:
```diff
@@ -0,0 +1,64 @@
+
+Previously, the queue didn't have any capacity limits. This resulted in a worker process exhausting all available memory if data was produced fast enough and the upstream server would be slow or unavailable.
+
+Now, you can configure the maximum number of entries that can be queued at any time with the `max_entries` queue parameter. When a queue reaches the maximum number of entries queued and another entry is enqueued, the oldest entry in the queue is deleted to make space for the new entry.
```
```suggestion
Now, you can configure the maximum number of entries that can be queued at any time with the `max_entries` parameter. When a queue reaches the maximum number of entries queued and another entry is enqueued, the oldest entry in the queue is deleted to make space for the new entry.
```

Guaris commented on app/_src/gateway/production/sizing-guidelines.md:
```diff
@@ -93,6 +93,43 @@ around **500MB** of memory allocated per worker process.
+
+Several plugins that are distributed with {{site.base_gateway}}
+utilize internal, in-memory queues to decouple production of data from
+the transmission to an upstream server.  These queues reduce the
```
```suggestion
the transmission to an upstream server. These queues reduce the
```

Guaris commented on app/_src/gateway/production/sizing-guidelines.md:
```diff
@@ -93,6 +93,43 @@ around **500MB** of memory allocated per worker process.
+the transmission to an upstream server.  These queues reduce the
+number of concurrent requests that are made to an upstream server
+under high load conditions and provide for buffering during temporary
+network and upstream outages.  Please see the [Plugin Queues]()
```
```suggestion
network and upstream outages. Please see the [Plugin Queues]()
```

Guaris replied:
No link here?

Guaris commented on app/_src/gateway/production/sizing-guidelines.md:
```diff
@@ -93,6 +93,43 @@ around **500MB** of memory allocated per worker process.
+configuration.
+
+The `queue.max_entries` configuration parameter determines how many
+entries can be waiting for transmission on a given queue.  Once this
```
```suggestion
entries can be waiting for transmission in a given queue. Once this
```

Guaris commented on app/_src/gateway/production/sizing-guidelines.md:
```diff
@@ -93,6 +93,43 @@ around **500MB** of memory allocated per worker process.
+
+In larger configurations, it is advisable to experimentally determine
+the memory requirements of queues by running {{site.base_gateway}} in
+a test environment and observe its memory consumption while plugin
```
```suggestion
a test environment and observing its memory consumption while plugin
```

Guaris commented on app/_src/gateway/production/sizing-guidelines.md:
```diff
@@ -93,6 +93,43 @@ around **500MB** of memory allocated per worker process.
+In larger configurations, it is advisable to experimentally determine
+the memory requirements of queues by running {{site.base_gateway}} in
+a test environment and observe its memory consumption while plugin
+upstream servers are unavailable, forcing queues to grow up to their
```
```suggestion
upstream servers are unavailable, forcing queues to reach the
```

Guaris reviewed (changes requested):
Small things only.

cloudjumpercat commented on app/_src/gateway/production/sizing-guidelines.md:
```diff
@@ -93,6 +93,43 @@ around **500MB** of memory allocated per worker process.
+the transmission to an upstream server. These queues reduce the
+number of concurrent requests that are made to an upstream server
+under high load conditions and provide for buffering during temporary
+network and upstream outages. Please see the [Plugin Queues]()
+documentation for details on {{site.base_gateway}}'s internal
+queueing system.
```
```suggestion
network and upstream outages. For more information about {{site.base_gateway}}'s internal queueing system, see [About Plugin Queuing](/gateway/{{page.kong_version}}/kong-plugins/queue/).
```

cloudjumpercat commented on app/_src/gateway/production/sizing-guidelines.md:
```diff
@@ -93,6 +93,43 @@ around **500MB** of memory allocated per worker process.
 
 For example, on a machine with 4 CPU cores and 8 GB of RAM available, we recommend allocating between 4-6 GB to cache via the `mem_cache_size` directive, depending on what other processes are running alongside {{site.base_gateway}}.
 
+### Plugin queues
```
```suggestion
{% if_version gte:3.3.x %}
### Plugin queues
```

cloudjumpercat commented on app/_src/gateway/production/sizing-guidelines.md:
```diff
@@ -93,6 +93,43 @@ around **500MB** of memory allocated per worker process.
+
+The default value of 10,000 entries for the `queue.max_entries` should
+provide for enough buffering in many installations while keeping the
+maximum memory usage of queues at reasonable levels.
```
```suggestion
maximum memory usage of queues at reasonable levels.
{% endif_version %}
```

cloudjumpercat replied:
```suggestion
maximum memory usage of queues at reasonable levels.
{% endif_version %}
```

cloudjumpercat replied:
```suggestion
maximum memory usage of queues at reasonable levels.
{% endif_version %}
```

cloudjumpercat reviewed (approved):
Okay, everything LGTM now! Thanks for all of your hard work on this @hanshuebner !
